### PR TITLE
[276] Add admin / rep suite assignment to draws

### LINKED ISSUE #276: Admins/reps can assign suites to groups in draws with lottery numbers
Fix the policy when we write the policy in #210

### PR DESCRIPTION
**Pull request by orenyk**
Resolves #276
- Add BulkSuiteSelector service
- Add :select_suites and :assign_suites routes to draw

Not a huge fan of radio buttons vs dropdowns, but we can discuss. I left the testing code in for both - we'll obviously want to clean that up before merging.

### CODE REVIEW

File: app/services/bulk_suite_selector.rb [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    selector.errors
+  end
+
+  def all_groups_selected
+    return if all_groups_selected?
+    errors.add(:base, 'You must select a suite for all groups')
+  end
+
+  def all_groups_selected?
+    params.select { |_k, v| !v.empty? }.count == groups.count
```
Comment (esoterik) [inline app/services/bulk_suite_selector.rb]: `params.values.reject(&:empty?).count == groups.count` is more concise

File: app/services/bulk_suite_selector.rb [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  private
+
+  attr_reader :groups, :params
+
+  # note that this occurs in a transaction
+  def process_all_suite_selections
+    results = groups.each_with_object({}) do |g, hash|
+      hash[g] = process_suite_selection(g)
+    end
+    return unless results.values.any? { |r| !r.empty? }
```
Comment (esoterik) [inline app/services/bulk_suite_selector.rb]: `return unless results.values.none?(&:empty?)`

File: app/services/bulk_suite_selector.rb [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    return if params.values.uniq == params.values
+    errors.add(:base, 'You must select different suites for each group')
+  end
+
+  def suites_in_correct_draw
+    return unless all_groups_selected?
+    return unless groups.any? do |g|
+      next false unless g.draw
+      suite = Suite.find_by(id: send(param_for(g)))
+      next false if suite&.draws&.include? g.draw
```
Comment (esoterik) [inline app/services/bulk_suite_selector.rb]: I don't think I like the use of `&.` here -- can we do `g.draw.suites.available.include? suite` instead?

File: app/services/bulk_suite_selector.rb [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      true
+    end
+    errors.add(:base, 'All groups in draws must be assigned to suites in '\
+      'the same draw')
+  end
+
+  def errors_from_results(results_hash)
+    results_hash.each_with_object([]) do |(g, r), array|
```
Comment (esoterik) [inline app/services/bulk_suite_selector.rb]: is there a less-complicated way to store these errors? 

if the refactoring would take too much time, we should at least rename `r` to `result` to clarify things

Comment (esoterik) [inline app/services/bulk_suite_selector.rb]: i guess we should also question if it's necessary to actually display all of the individual errors?

i'm sort of okay with a generic "oops something went wrong in assigning {GROUP NAME}" for errors that should not be reachable from the frontend, though that might not be the best

File: app/services/bulk_suite_selector.rb [hunk 5]
```diff
@@ -0,0 +1,136 @@
+# frozen_string_literal: true
+#
+# Service object to handle the bulk assignment of suites to groups. Ensures that
+# suites are available, groups don't have suites assigned, and optionally that
+# suites belong to the appropriate draw.
+class BulkSuiteSelector
```
Comment (esoterik) [inline app/services/bulk_suite_selector.rb]: should this be a form object? i think all/most of the logic for that is already here, unless i'm missing something

File: app/services/suite_selector.rb [hunk 6]
```diff
@@ -34,7 +36,7 @@ def select
 
   private
 
-  attr_accessor :errors
+  attr_writer :errors
```
Comment (esoterik) [inline app/services/suite_selector.rb]: why was this necessary? can it just stay `attr_accessor`?

File: app/policies/draw_policy.rb [hunk 7]
```diff
@@ -86,6 +86,14 @@ def start_selection?
     edit? && record.lottery? && record.lottery_complete?
   end
 
+  def select_suites?
+    record.suite_selection? && (user.admin? || user.rep?)
```
Comment (esoterik) [inline app/policies/draw_policy.rb]: `!user.student?`?

File: app/services/bulk_suite_selector.rb (hunk 2 above)
Comment (orenyk) [inline app/services/bulk_suite_selector.rb]: Ooh nice

File: app/services/bulk_suite_selector.rb (hunk 5 above)
Comment (orenyk) [inline app/services/bulk_suite_selector.rb]: You're right, it basically is a form object. I'll rename / move it as appropriate.

File: app/services/bulk_suite_selector.rb (hunk 1 above)
Comment (orenyk) [inline app/services/bulk_suite_selector.rb]: Needs to be `reject!` but good suggestion.

File: app/services/bulk_suite_selector.rb (hunk 3 above)
Comment (orenyk) [inline app/services/bulk_suite_selector.rb]: Much better.

File: app/services/bulk_suite_selector.rb (hunk 4 above)
Comment (orenyk) [inline app/services/bulk_suite_selector.rb]: These errors are kinda reachable from the front-end, and I'd rather be more verbose now and deal with it down the road when we figure out the frontend properly. I renamed the block variables for clarity.

File: app/services/suite_selector.rb (hunk 6 above)
Comment (orenyk) [inline app/services/suite_selector.rb]: I believe that this overwrites the public reader (if I'm remembering correctly).

File: app/policies/draw_policy.rb (hunk 7 above)
Comment (orenyk) [inline app/policies/draw_policy.rb]: I'd rather be explicit with permissions than not.

**Comment by orenyk** [comment]
@esoterik updated!

**Comment by orenyk** [comment]
Not sure what you mean, let's discuss tomorrow.
